cpp/examples/arrow/dataset-parquet-scan-example.cc

```diff
@@ -0,0 +1,136 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdlib>
+#include <iostream>
+
+#include <arrow/api.h>
+
+#include <arrow/dataset/dataset.h>
+#include <arrow/dataset/discovery.h>
+#include <arrow/dataset/file_base.h>
+#include <arrow/dataset/file_parquet.h>
+#include <arrow/dataset/filter.h>
+#include <arrow/dataset/scanner.h>
+#include <arrow/filesystem/localfs.h>
+#include <arrow/util/iterator.h>
+#include <arrow/util/task_group.h>
+#include <arrow/util/thread_pool.h>
+
+using arrow::field;
+using arrow::int16;
+using arrow::Schema;
+using arrow::Table;
+
+using arrow::internal::TaskGroup;
+using arrow::internal::ThreadPool;
+
+using arrow::fs::FileSystem;
+using arrow::fs::LocalFileSystem;
+using arrow::fs::Selector;
+
+using arrow::dataset::Dataset;
+using arrow::dataset::DataSource;
+using arrow::dataset::DataSourceDiscovery;
+using arrow::dataset::DataSourceVector;
+using arrow::dataset::Expression;
+using arrow::dataset::FieldExpression;
+using arrow::dataset::FileFormat;
+using arrow::dataset::FileSystemDataSourceDiscovery;
+using arrow::dataset::ParquetFileFormat;
+using arrow::dataset::PartitionScheme;
+using arrow::dataset::Scanner;
+using arrow::dataset::ScannerBuilder;
+using arrow::dataset::ScanTask;
+
+#define ABORT_ON_FAILURE(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      abort();                                     \
+    }                                              \
+  } while (0);
+
+std::shared_ptr<Dataset> GetDatasetFromPath(FileSystem* fs,
+                                            std::shared_ptr<FileFormat> format,
+                                            std::string path) {
+  // Find all files under `path`
+  Selector s;
+  s.base_dir = path;
+  s.recursive = true;
+
+  std::shared_ptr<DataSourceDiscovery> discovery;
+  ABORT_ON_FAILURE(FileSystemDataSourceDiscovery::Make(fs, s, format, &discovery));
+
+  std::shared_ptr<Schema> inspect_schema;
+  ABORT_ON_FAILURE(discovery->Inspect(&inspect_schema));
+
+  std::shared_ptr<DataSource> source;
+  ABORT_ON_FAILURE(discovery->Finish(&source));
+
+  return std::make_shared<Dataset>(DataSourceVector{source}, inspect_schema);
+}
+
+std::shared_ptr<Scanner> GetScannerFromDataset(std::shared_ptr<Dataset> dataset,
+                                               std::vector<std::string> columns,
+                                               std::shared_ptr<Expression> filter) {
+  std::unique_ptr<ScannerBuilder> scanner_builder;
+  ABORT_ON_FAILURE(dataset->NewScan(&scanner_builder));
+
+  if (!columns.empty()) {
+    ABORT_ON_FAILURE(scanner_builder->Project(columns));
+  }
+
+  if (filter != nullptr) {
```

Comment:
nit: filter should never be nullptr

Reply:
I made it such that because I meant the example to have options in the future, e.g. take the filter from argv (--filter=100, --repeat-source=3, ...).

Reply:
This is also to show that Filter() is optional, I'll leave it like this.

Reply:
If filter is optional, then please make the default `scalar(true)`